api/osb/reference_instance_plugin.go

```diff
@@ -0,0 +1,361 @@
+
+	if err != nil {
+		return nil, err
+	}
+	if !isReferencePlan {
+		return next.Handle(req)
+	}
+
+	return p.generateOSBResponse(ctx, "FetchService", instance)
+}
+
+func (p *referenceInstancePlugin) generateOSBResponse(ctx context.Context, method string, instance *types.ServiceInstance) (*web.Response, error) {
+	var marshal []byte
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json")
+	switch method {
+	case "Provision":
+		return &web.Response{
+			Body:       []byte(`{}`),
+			StatusCode: http.StatusCreated,
+			Header:     headers,
+		}, nil
+	case "Deprovision":
+		return &web.Response{
+			Body:       []byte(`{}`),
+			StatusCode: http.StatusOK,
+			Header:     headers,
+		}, nil
+	case "UpdateService":
+		return &web.Response{
+			Body:       []byte(`{}`),
+			StatusCode: http.StatusOK,
+			Header:     headers,
+		}, nil
+	case "FetchService":
+		osbResponse, err := p.buildOSBFetchServiceResponse(ctx, instance)
+		if err != nil {
+			return nil, err
+		}
+		marshal, err = json.Marshal(osbResponse)
+		if err != nil {
+			return nil, err
+		}
+		return &web.Response{
+			Body:       marshal,
+			StatusCode: http.StatusOK,
+			Header:     headers,
+		}, nil
+	}
+	return nil, util.HandleInstanceSharingError(util.ErrUnknownOSBMethod, method)
+
+}
+
+func (p *referenceInstancePlugin) buildOSBFetchServiceResponse(ctx context.Context, instance *types.ServiceInstance) (osbObject, error) {
+	serviceOffering, plan, err := p.getServiceOfferingAndPlanByPlanID(ctx, instance.ServicePlanID)
+	if err != nil {
+		return nil, util.HandleInstanceSharingError(util.ErrNotFoundInStorage, string(types.ServiceOfferingType))
+	}
+	osbResponse := osbObject{
+		"service_id":       serviceOffering.CatalogID,
+		"plan_id":          plan.CatalogID,
+		"maintenance_info": instance.MaintenanceInfo,
+		"parameters": map[string]string{
+			constant.ReferencedInstanceIDKey: instance.ReferencedInstanceID,
+		},
+	}
+	return osbResponse, nil
+}
+
+// PollBinding intercepts poll binding operation requests and check if the instance is in the platform from where the request comes
+func (p *referenceInstancePlugin) PollBinding(req *web.Request, next web.Handler) (*web.Response, error) {
+	return p.handleBinding(req, next)
+}
+
+func (p *referenceInstancePlugin) validateOwnership(req *web.Request) error {
+	ctx := req.Context()
+	callerTenantID := gjson.GetBytes(req.Body, "context."+p.tenantIdentifier).String()
+	path := fmt.Sprintf("parameters.%s", constant.ReferencedInstanceIDKey)
+	referencedInstanceID := gjson.GetBytes(req.Body, path).String()
+	byID := query.ByField(query.EqualsOperator, "id", referencedInstanceID)
+	dbReferencedObject, err := p.repository.Get(ctx, types.ServiceInstanceType, byID)
+	if err != nil {
+		return util.HandleStorageError(err, types.ServiceInstanceType.String())
+	}
+	instance := dbReferencedObject.(*types.ServiceInstance)
+	referencedOwnerTenantID := instance.Labels["tenant"][0]
```

Comment:
do we have a test for non multi-tenant flow?

Reply:
No, we will have to cover it. 
I keep it open for now.